### PR TITLE
Scale image with preserved aspect ratio

### LINKED ISSUE #1952: Options for max width/height or keep aspect ratio for Camera
I have a feature request about adding options to the Camera plugin for setting max `width`/`height`, alternatively a new option for keeping aspect ratio (e.g. `keepRatio`).

This is somewhat related to #217, in which @mlynch wrote:
> _Just added this feature to master and will release tomorrow. you can set `width` and/or `height`. If you set one but not the other, it does an aspect ratio scale to fit that dimension._

What I need is a further extension of the scaling capabilities, I want to be able to limit the _longest_ side of the photo, yet keep the aspect ratio. I've used the [Cordova Camera plugin](https://cordova.apache.org/docs/en/latest/reference/cordova-plugin-camera/index.html) before, in which this was possible.

The problem now is that
- if I specify both `width` _and_ `height`, I'm sure no side is too long, but then the aspect ratio of the resulting image is changed.
- if I specify only `width` _or_ `height`, the aspect ratio is retained, but then I don't know if my prescribed value is applied to the longer or shorter side of the image, since I don't know beforehand if the image is in portrait or landscape format.

I don't know if it's good that `width` and `height` work like in the Cordova case, always retaining aspect ratio, or if it's better to have an additional option enforcing that aspect ratio is kept (e.g. `keepRatio`). The latter alternative gives more flexibility, but maybe makes things too complicated.

Let me know what you think of this!
Is this a feature that can be added?

### PR DESCRIPTION
**Pull request by sandstrom**
When resizing images it makes sense to restrain width/height, but preserve aspect ratio this is how the camera plugin with how the Cordova camera plugin behave[1], and all other resize scenarios I've come across. 

So I hope this is seen as a pretty straight forward enhancement, even though it's technically breaking in case someone wanted to have their images squeezed (unlikely).

If you want to preserve backwards compatibility you could add a fourth argument called `preserveAspectRatio` and let it be false by default. The flip the flag (or drop it) in the next major version.

***
These are the first few lines of Java that I've written in my entire life, so there are likely some bugs in it. I don't have a java environment on my machine, so great if someone else could carry this over the finish line for me. 😄 

This is blocking our migration from Cordova to Capacitor so if someone would like to take a look at it it would be most appreciated 🙏 

@mlynch and @jcesarmobile I saw you worked on something similar in https://github.com/ionic-team/capacitor/issues/217, so perhaps you'd like to take a look at this and let me know what you think?

Fixes #1952

[1] https://cordova.apache.org/docs/en/latest/reference/cordova-plugin-camera/#module_camera.CameraOptions

### CODE REVIEW

**Comment by PR1YANKPAT3L** [comment]
[Line 55](https://github.com/ionic-team/capacitor/blob/d063d79c2b4ff882c4d517498d720eb92d7ad2c2/android/capacitor/src/main/java/com/getcapacitor/plugin/camera/ImageUtils.java#L55) should be bitmap instead of Bitmap

**Comment by sandstrom** [comment]
@PR1YANKPAT3L Thanks! 😄

**Comment by sandstrom** [comment]
@jcesarmobile I've fixed the failed test now, please have a look when you have time! 😄

File: android/capacitor/src/main/java/com/getcapacitor/plugin/camera/ImageUtils.java [hunk 1]
```diff
@@ -19,24 +19,38 @@
...
+    int newWidth = Math.min(width, maxWidth);
+    int newHeight = (height * newWidth) / width;
+
+    if (newHeight > maxHeight) {
+      newWidth = (width * maxHeight) / height;
+      newHeight = maxHeight;
     }
 
-    return bitmap;
+    newWidth = Math.round(newWidth);
```
Comment (jcesarmobile) [inline android/capacitor/src/main/java/com/getcapacitor/plugin/camera/ImageUtils.java]: They are `int` already, so round won't do anything on them.
But `int` just removes everything after the `.`.  (i.e. 2.9 becomes 2, not 3).
So you probably need to use float for the variables where you put a value that is divided, and then use round on the floats to make them int.

File: android/capacitor/src/main/java/com/getcapacitor/plugin/camera/ImageUtils.java [hunk 2]
```diff
@@ -19,24 +19,38 @@
...
+    newHeight = Math.round(newHeight);
+
+    if (newWidth == width && newHeight == height) {
```
[the remark this answers is not in the document]
Comment (sandstrom) [reply android/capacitor/src/main/java/com/getcapacitor/plugin/camera/ImageUtils.java]: Thanks, fixed this!

**Comment by sandstrom** [comment]
@jcesarmobile I've never worked with iOS or Android native development, so if you'd like to take over from here and bring it over the finish line that would be awesome! 🏁

**Comment by sandstrom** [comment]
@jcesarmobile Would you like to take a look at this?

This PR is (hopefully) ~50% complete, great if you'd like to take it over the finish line. I don't have enough experience with swift/java to finish this unfortunately.

**Comment by priyankpat** [comment]
I can take a look at this once I have some time.

**Comment by sandstrom** [comment]
@priyankpat Do you have time to look at this? Would be awesome! 😄

**Comment by priyankpat** [comment]
Sorry have been busy with work lately, I will have an update in a day or two.

**Comment by sandstrom** [comment]
friendly ping @priyankpat 😄

**Comment by sandstrom** [comment]
ping @priyankpat 😄

**Comment by iphilgood** [comment]
Hey @sandstrom 

I'm having an iOS version ready. I'm not quite sure what my next steps are? I've cloned your repo and would like to commit the change. Do you need to add me as a contributor? 😅

**Comment by priyankpat** [comment]
My xcode is broken 😢 , please give this a try while I try to make it work.

```swift
func resizeImage(_ image: UIImage) -> UIImage? {
    let width: Int = image.width
    let height: Int = image.height
    
    let desiredMaxWidth: Int = settings.width
    let desiredMaxHeight: Int = settings.height
    
    // 0 is treated as 'no restriction'
    let maxWidth: Int = desiredMaxWidth == 0 ? width : desiredMaxWidth
    let maxHeight: Int = desiredMaxHeight == 0 ? height : desiredMaxHeight
    
    // resize with preserved aspect ratio
    let newWidth: Float = min(width, maxWidth)
    let newHeight: Float = (height * newWidth) / width
    
    if newHeight > maxHeight {
        newWidth = (width * maxHeight) / height
        newHeight = maxHeight
    }
    
    var size = CGSize.init(width: Int(newWidth), height: Int(newHeight))

    UIGraphicsBeginImageContextWithOptions(size, false, 1.0)
    image.draw(in: CGRect(origin: CGPoint.zero, size: size))

    let scaledImage = UIGraphicsGetImageFromCurrentImageContext()
    UIGraphicsEndImageContext()
    return scaledImage
  }
```

**Comment by iphilgood** [comment]
@priyankpat Did you already update to Big Sur and the Xcode Beta? 😂 I'm gonna try that. But what would be my next step? I'm not quite sure how I can contribute to this 😅

**Comment by iphilgood** [comment]
@priyankpat the code failed because of the typings. I've fixed it. What do you think about that?

```swift
func resizeImage(_ image: UIImage) -> UIImage? {
  let imageHeight = Float(image.size.height)
  let imageWidth = Float(image.size.width)

  // 0 is treated as 'no restriction'
  let maxHeight = settings.height == 0 ? imageHeight : settings.height
  let maxWidth = settings.width == 0 ? imageWidth : settings.width

  // resize with preserved aspect ratio
  var newWidth = min(imageWidth, maxWidth)
  var newHeight = (imageHeight * newWidth) / imageWidth
  if (newHeight > maxHeight) {
    newWidth = (imageWidth * maxHeight) / imageHeight;
    newHeight = maxHeight;
  }
  let size = CGSize.init(width: Int(newWidth), height: Int(newHeight))

  UIGraphicsBeginImageContextWithOptions(size, false, 1.0)
  image.draw(in: CGRect(origin: CGPoint.zero, size: size))

  let scaledImage = UIGraphicsGetImageFromCurrentImageContext()
  UIGraphicsEndImageContext()
  return scaledImage
}
```

**Comment by dwieeb** [comment]
Given that this is either going to be a breaking change or require adding a `preserveAspectRatio` option, I added an issue for the Camera plugin for Capacitor 3: https://github.com/ionic-team/capacitor-plugins/issues/2

We want to limit new features for Capacitor 2 at this time, and with what we're planning for the plugins (see https://github.com/ionic-team/capacitor/issues/3227), it just makes sense to address this later. This will allow us to make a breaking change in behavior to preserve the aspect ratio by default.

Thanks for the PR!

**Comment by firthy** [comment]
I as many others probably need this feature now. At the moment i can't imagine anyone wanting to resize an image to have the resulting image distorted. Why not add the preserveAspectRatio as false and then remove it in the next version. At least the resize will behave as expected (not distort the image). I have a project that is now in limbo due to this issue. Waiting for the next version is not really an option.

**Comment by dwieeb** [comment]
Anyone is welcome to submit a full implementation of this feature that's easy for the team to test, but as it stands right now, there are other things to prioritize. We need to start working on Capacitor 3.

**Comment by iphilgood** [comment]
@dwieeb where exactly should I submit a PR? Is it here or over at the separate plugin repo?

**Comment by dwieeb** [comment]
@iphilgood Here for now!

**Comment by sandstrom** [comment]
@dwieeb This is not a breaking change, this is a bug. It'll never make any sense to *not* preserve the aspect ratio.

**Comment by dwieeb** [comment]
I can understand that point of view, and I agree it doesn't make sense to not preserve aspect ratio when resizing photos from the Camera, but you might be surprised by what some people consider a breaking change. We're just trying to be careful not to break existing functionality.

**Comment by iphilgood** [comment]
@dwieeb @sandstrom i've created my first PR over here #3297 👍

**Pull request closed without merging**